src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -182,14 +211,26 @@ public bool IsConnected
         /// <summary>Set to <c>1</c> if the peer disconnection has been initiated, <c>0</c> otherwise.</summary> 
         private int disconnected;
 
+        /// <summary>Set to <c>1</c> if the peer disposal has been initiated, <c>0</c> otherwise.</summary> 
+        private int disposed;
+        
+        /// <summary>
+        /// Async context to allow to recognize whether <see cref="onDisconnected"/> callback execution is scheduled in this async context.
+        /// <para>
+        /// It is not <c>null</c> if one of the following callbacks is in progress: <see cref="StateChanged"/>, <see cref="MessageReceived"/>,
+        /// set to <c>null</c> otherwise.
+        /// </para>
+        /// </summary>
+        private readonly AsyncLocal<DisconnectedExecutionAsyncContext> onDisconnectedAsyncContext;
```

Comment:
hmmm... Is this absolutely necessary, we are introducing more levels of complexity here, I assume this will follow a message during its propagation inside the behaviors and that way we can know if disconnect was called from a behaviour or not.

Reply:
Yes. Since we want to prevent callback call from inside of another callback this is needed.

Reply:
Alternative solution would be to have a context parameter that everyone would be required to pass to every sub- function. That doesn't seem better at all. Do you see any other alternative?